[flutter_tools] Cleanup of native asset related code (removes around 50% of the native asset related code)

mkustermann opened the pull request:
tl;dr Removes 50% (>1650 locs) of native asset related code in `packages/flutter_tools`

Before this PR the invocation of dart build/link/dry-run was implemented per OS. This lead to very large code duplication of almost identical, but sligthly different code. It also led to similarly duplicated test code.

Almost the entire dart build/link/dry-run implementation is identical across OSes. There's small variations:

  - configuration of the build (e.g. android/macos/ios version, ios sdk, ...)
  - determining target locations & copying the final shared libraries

This PR unifies the implementation by reducing the code to basically two main functions:

  * `runFlutterSpecificDartBuild` which is responsible for
    - obtain flutter configuration
    - perform dart build (& link)
    - determine target location & install binaries

  * `runFlutterSpecificDartDryRunOnPlatforms` which is responsible for a similar (but not same):
    - obtain flutter configuration
    - perform dart dry run
    - determine target location

these two functions will call out to helpers for the OS specific functionality:

  * `_assetTargetLocationsForOS` for determining the location of the code assets

  * `_copyNativeCodeAssetsForOS` for copying the code assets (and possibly overriting the install name, etc)

=> Since we get rid of the code duplication across OSes and have only a single code path for the build/link/dry-run, we can also remove the duplicated tests that were pretty much identical across OSes.

We also harden the building code by adding asserts, e.g.

  * the dry fun functionality should never be used by `flutter test`

  * the `build/native_assets/<os>/native_assets.yaml` should only be used by `flutter test` and the dry-run of `flutter run`

=> We change the tests to also comply with these invariants (so the tests are not testing things that cannot happen in reality)

We also rename `{,Flutter}NativeAssetsBuildRunner` to disambiguate it from the `package:native_asset_builder`'s `NativeAssetsBuildRunner`.

We also reorganize the main code to make it readable from top-down and make members private where they can be.

mkustermann commented:
/cc @mosuem That's preparation for adding support for data assets (which would be much harder without cleaning up this code first)

dcharkes commented on packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:
```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
+  switch (targetOS) {
+    case OS.windows:
+    case OS.linux:
+      await _copyNativeCodeAssetsToBundleOnWindowsLinux(
```
Maybe assert that `codesignIdentity` is null. Since we don't do code signing. If we ever add code signing, we should not forget to use it here.

mkustermann replied:
Done - though I've had an assert in another place already.

dcharkes commented on packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:
```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
 
-class HotRunnerNativeAssetsBuilderImpl implements HotRunnerNativeAssetsBuilder {
-  const HotRunnerNativeAssetsBuilderImpl();
+/// With `flutter run -d all` we need a place to store the native assets
```
Nit: The diff is larger than it needs to be due to reorderings. This makes reviewing harder and redirects questions to you via git blame. If you feel strongly about the order of code, fine be me.

dcharkes commented on packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:
```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
-      buildRunner: buildRunner,
-      flutterDevices: flutterDevices,
-    );
+Map<NativeCodeAssetImpl, KernelAsset> _assetTargetLocationsWindowsLinux(
```
The previous reviewers preferred not having OS-specific things in this OS-agnostic file. Please rename back to `_assetTargetLocationsSingleArchitecture` to comply with their wishes.

And ditto for code moving.

mkustermann replied:
If we have `_assetTargetLocations{Android,IOS,MacOS}` a reader of this code would naturally look for `_assetTargetLocations{Linux,Windows}` and cannot find them - which is the first confusing.

If we name it `_assetTargetLocationsSingleArchitecture`, then one could think that `_assetTargetLocationsAndroid`  means it's always multiple architectures, but it's not necessarily the case.

The difference between these functions isn't whether it's a single architecture or multiple but which operating system they deal with. So I think it makes more sense to have this naming.

@dcharkes If you feel strongly about this, I can rename it back, but I think it's worse for a reader of this code.

dcharkes replied:
> @dcharkes If you feel strongly about this, I can rename it back, but I think it's worse for a reader of this code.

I don't, I had `LinuxWindows` in my initial PRs as well before the reviewers on those PRs requested changes.

mkustermann replied:
Then I'll leave it as it is.

dcharkes commented on packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:
```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
-}) async {
-  if (await nativeBuildRequired(buildRunner)) {
-    return null;
+Future<void> _copyNativeCodeAssetsToBundleOnWindowsLinux(
```
Ditto, avoid using OSspecific names in this file.

Ditto, shows up as diff even though didn't change.

mkustermann replied:
It makes consistent naming with all the other `_copy*` methods.

I've tried to mimize the diff now as much as I could.

dcharkes commented on packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:
```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
+      : await buildRunner.cCompilerConfig;
+  final IOSSdkImpl? targetIOSSdk =
+      targetOS == OS.iOS ? getIOSSdk(environmentType!) : null;
+  // TODO(dcharkes): Fetch minimum iOS version from somewhere. https://github.com/flutter/flutter/issues/145104
```
OS-specific things should live in the OS-specific files. Maybe move helper functions there (also for targetIOSVersion and targetMacOSVersion) and leave a comment on the `buildRunner.build` invocation it contains the config for all OSes. And then on the helper functions that they return null for the irrelevant OSes?

mkustermann replied:
Done.

dcharkes commented on packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart:
```diff
@@ -99,13 +100,13 @@ Future<void> setInstallNamesDylib(
       newInstallName,
-      for (final MapEntry<String, String> entry in oldToNewInstallNames.entries)
-        ...<String>['-change', entry.key, entry.value],
+      for (final MapEntry<String, String> entry in oldToNewInstallNames
```
I hate to do this. But it looks like the formatter ran and made a bunch of whitespace changes. (Also in other files.)

I will refer to the Flutter team on if we care about that in the days before turning the formatter on here in the codebase.

mkustermann replied:
I've tried to manually revert the auto-formatted lines.

dcharkes commented on packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart:
```diff
@@ -135,6 +135,7 @@ void main() {
               const Barrier('Application finished.'),
             ],
             logging: false,
+            debug: true,
```
Revert this?

dcharkes reviewed (approved):
LGTM!

Thanks @mkustermann! 🚀 

I'll refer to the Flutter team on unrelated code having been formatted by the formatter.

dcharkes commented:
Probably @christopherfujino or @andrewkolos should review this. (Or @bkonyi.)

mkustermann committed:
[flutter_tools] Cleanup of native asset related code

tl;dr Removes 50% (>1650 locs) of native asset related code in
      `packages/flutter_tools`

Before this PR the invocation of dart build/link/dry-run was implemented
per OS. This lead to very large code duplication of almost identical,
but sligthly different code. It also led to similarly duplicated test
code.

Almost the entire dart build/link/dry-run implementation is identical
across OSes. There's small variations:

  - configuration of the build (e.g. android/macos/ios version, ios sdk, ...)
  - determining target locations & copying the final shared libraries

This PR unifies the implementation by reducing the code to basically two
main functions:

  * `runFlutterSpecificDartBuild` which is responsible for
      - obtain flutter configuration
      - perform dart build (& link)
      - determine target location & install binaries

  * `runFlutterSpecificDartDryRunBuild`

these two functions will call out to helpers for the OS specific
functionality:

  * `_assetTargetLocationsForOS` for determining the location of the
    code assets

  * `_copyNativeCodeAssetsForOS` for copying the code assets (and
    possibly overriting the install name, etc)

=> Since we get rid of the code duplication across OSes and have only a
single code path for the build/link/dry-run, we can also remove the
duplicated tests that were pretty much identical across OSes.

We also harden the building code by adding asserts, e.g.

  * the dry fun functionality should never be used by `flutter test`

  * the `build/native_assets/<os>/native_assets.yaml` should only be used
    by `flutter test` and the dry-run of `flutter run`

=> We change the tests to also comply with these invariants (so the
tests are not testing things that cannot happen in reality)

We also rename `{,Flutter}NativeAssetsBuildRunner` to disambiguate it
from the `package:native_asset_builder`'s `NativeAssetsBuildRunner`.

mkustermann commented:
@dcharkes I've rebased, tried to align old/new code to make diff smaller, tried to un-auto-format code. Please let me know if there's something else.

@bkonyi This is mainly a refactoring of native asset code (for which I'd consider @dcharkes the most knowledgeable person / owner) - do you also want to review this?

dcharkes commented on packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart:
```diff
@@ -271,7 +271,7 @@ void main() {
             buildSubcommand,
             if (buildSubcommand == 'ios') '--no-codesign',
-            '-v', // Requires verbose mode for error.
+            '-v' // Requires verbose mode for error.
```
formatting merge change?

mkustermann replied:
Done.

dcharkes reviewed (approved):
lgtm from my side!

mkustermann committed:
add trailing comma again

bkonyi commented on packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart:
```diff
@@ -99,13 +99,13 @@ Future<void> setInstallNamesDylib(
       newInstallName,
-      for (final MapEntry<String, String> entry in oldToNewInstallNames.entries)
-        ...<String>['-change', entry.key, entry.value],
+      for (final MapEntry<String, String> entry in oldToNewInstallNames
```
Ubernit: consider using an object destructuring pattern to pull out the key and value.

mkustermann replied:
I think that makes it even more verbose, because of lints that require explicit type annotations (then for both a key and a value variable).

As this is just refactoring of existing code, I hope it's ok if I leave it like this.

bkonyi replied:
Yup, that's no problem, just figured I'd mention it. I forgot about those lints...

bkonyi commented on packages/flutter_tools/test/general.shard/isolated/android/native_assets_test.dart:
```diff
@@ -287,12 +123,16 @@ void main() {
+    await runFlutterSpecificDartBuild(
+      environmentDefines: <String, String>{
+        kBuildMode: BuildMode.debug.cliName,
+        kMinSdkVersion: 21.toString(),
```
Can this be a constant somewhere?

mkustermann replied:
Done.

bkonyi commented on packages/flutter_tools/test/general.shard/isolated/android/native_assets_test.dart:
```diff
@@ -308,16 +148,19 @@ void main() {
+      () => runFlutterSpecificDartBuild(
+        environmentDefines: <String, String>{
+          kBuildMode: BuildMode.debug.cliName,
+          kMinSdkVersion: 21.toString(),
```
Ditto

mkustermann replied:
Done.

bkonyi commented on packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart:
```diff
@@ -355,22 +186,25 @@ void main() {
+      await runFlutterSpecificDartBuild(
+        environmentDefines: <String, String>{
+          kBuildMode: buildMode.cliName,
+          kSdkRoot: '.../iPhone Simulator',
```
I'm not super familiar with this code, but is this supposed to be a path?  What is this value for?

mkustermann replied:
Yes this is supposed to be a path and the implementation greps for `iPhone Simulator` in the name. See existing code at `.../src/macos/xcode.dart:environmentTypeFromSdkroot`

bkonyi commented on packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:
```diff
@@ -30,11 +29,159 @@ import 'macos/native_assets.dart';
+/// If any of the dependencies change, then the Dart build should be performed
+/// again.
+final class DartBuildResult {
+  const DartBuildResult(this.codeAssets, this.dependencies);
```
Nit: prefer required named arguments.

mkustermann replied:
Done.

Though to be honest I think it makes the code worse: This class is only used internally within flutter tools. Making them named will just increase code size and reduce information density:

e.g. What used to be
```dart
return DartBuildResult(codeAssets, dependencies);
```
becomes now
```dart
return DartBuildResult(codeAssets: codeAssets, dependencies: dependencies);
```
why is the second one better than the first?

bkonyi commented on packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:
```diff
@@ -30,11 +29,159 @@ import 'macos/native_assets.dart';
+  final List<Uri> dependencies;
+}
+
+/// Invokes the build of all transitive Dart packages & prepares code assets to
```
Ubernit: s/&/and/

mkustermann replied:
Done

bkonyi commented on packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:
```diff
@@ -30,11 +29,159 @@ import 'macos/native_assets.dart';
 /// It enables mocking `package:native_assets_builder` package.
 /// It also enables mocking native toolchain discovery via [cCompilerConfig].
-abstract class NativeAssetsBuildRunner {
+abstract class FlutterNativeAssetsBuildRunner {
```
Nit: can this be an `interface class`?

mkustermann replied:
Done.

bkonyi commented on packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:
```diff
@@ -794,10 +602,240 @@ KernelAsset _targetLocationSingleArchitecture(
+      assert(codesignIdentity == null);
+      await copyNativeCodeAssetsAndroid(
+          buildUri, assetTargetLocations, fileSystem);
+    default:
```
This isn't defined in flutter_tools, but is there any reason we aren't just using enhanced enums for `OS` and `OSImpl`? This would allow for us to perform a static exhaustiveness check for all these switches without requiring us to manually raise a `StateError` everywhere.

bkonyi replied:
Fair enough. Should we maybe be throwing `UnimplementedError` instead then?

mkustermann replied:
What Daco says. Adding more entries to the enum shifts from being a compile-time error to being a runtime error (and only if the new entry is used)

mkustermann replied:
Done.

bkonyi reviewed (approved):
LGTM with some nits.

mkustermann committed:
address comments

mkustermann commented:
Thanks, @bkonyi !

mkustermann committed:
Change StateError -> UnimplementedError

eyebrowsoffire commented:
Reason for revert: Postsubmit failures closing the tree. See the following examples:

https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20native_assets_ios/5738/overview
https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_arm64_mokey%20native_assets_android/583/overview
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_pixel_7pro%20native_assets_android/4075/overview
https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20native_assets_ios_simulator/5287/overview

mkustermann commented:
> Reason for revert: Postsubmit failures closing the tree. See the following examples:

Can reproduce the failure of the test, it was missing a case for iOS/Android, fix is
```diff
diff --git a/packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart b/packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart
index b28543ba16..fd8952e7d9 100644
--- a/packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart
+++ b/packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart
@@ -145,7 +145,9 @@ Future<Uri?> runFlutterSpecificDartDryRunOnPlatforms({
     final OSImpl targetOS = _getNativeOSFromTargetPlatfrorm(targetPlatform);
     if (targetOS != OS.macOS &&
         targetOS != OS.windows &&
-        targetOS != OS.linux) {
+        targetOS != OS.linux &&
+        targetOS != OS.android &&
+        targetOS != OS.iOS) {
       await ensureNoNativeAssetsOrOsIsSupported(
         projectUri,
```

Make a reland CL in https://github.com/flutter/flutter/pull/155745
